resources/User.py

```diff
@@ -1,63 +1,67 @@
 
-    # Login function: allows a user to login using their email and password as credentials
-    # The password is compared to the hashed password stored in the users table
-    # Once the password is verified, an API key is generated, which is stored in the users table and sent to the verified user
     def get(self):
+        """
+        Login function: allows a user to login using their email and password as credentials
+        The password is compared to the hashed password stored in the users table
+        Once the password is verified, an API key is generated, which is stored in the users table and sent to the verified user
+        """
         try:
             data = request.get_json()
             email = data.get("email")
             password = data.get("password")
             cursor = self.psycopg2_connection.cursor()
-            cursor.execute(
-                f"select id, password_digest from users where email = '{email}'"
-            )
-            results = cursor.fetchall()
-            user_data = {}
-            if len(results) > 0:
-                user_data = {"id": results[0][0], "password_digest": results[0][1]}
-            else:
-                return {"error": "no match"}
+
+            user_data = user_get_results(cursor, email)
             if check_password_hash(user_data["password_digest"], password):
-                api_key = uuid.uuid4().hex
-                user_id = str(user_data["id"])
-                cursor.execute(
-                    "UPDATE users SET api_key = %s WHERE id = %s", (api_key, user_id)
-                )
-                payload = {"api_key": api_key}
-                self.psycopg2_connection.commit()
-                return payload
+                return {"data": "Successfully logged in"}
 
         except Exception as e:
             self.psycopg2_connection.rollback()
             print(str(e))
             return {"error": str(e)}
 
-    # Sign up function: allows a user to sign up by submitting an email and password. The email and a hashed password are stored in the users table and this data is returned to the user upon completion
     def post(self):
+        """
+        Sign up function: allows a user to sign up by submitting an email and password.
+        The email and a hashed password are stored in the users table and this data is returned to the user upon completion
+        """
+        try:
+            data = request.get_json()
+            email = data.get("email")
+            password = data.get("password")
+            cursor = self.psycopg2_connection.cursor()
+            user_post_results(cursor, email, password)
+            self.psycopg2_connection.commit()
+
+            return {"data": "Successfully added user"}
+
+        except Exception as e:
+            self.psycopg2_connection.rollback()
+            print(str(e))
+            return {"error": e}
+
+    # Endpoint for updating a user's password
+    def put(self):
```

Comment:
@mbodeantor In the case of resetting a user's password, how are we authenticating them? Given the structure of this endpoint, anyone with access could change anyone else's password.

I'm thinking we might want to send the user an email with a magic link, and only by clicking that email are they able to access the frontend UI for changing a password. Wyt?

Reply:
@joshuagraber That makes sense to me. What API changes would be required for that?

Reply:
I'm on board except I think instead of assuming in point 2, can we have a "forgot password?" link that only requires them to send an email?

Reply:
I still think @mbodeantor should answer, but I'll note here that we could consider adding an external library so we don't reinvent any wheels:

- [flask-security-too](https://flask-security-too.readthedocs.io/)
- [authlib](https://authlib.org/)
- [flask-user](https://flask-user.readthedocs.io/en/latest/) + [flask-login](https://pypi.org/project/Flask-Login/)

Reply:
For sure @mbodeantor I'm thinking that there can be a "change password" link (enter current pw + desired updated pw) and a "reset password" link (only email). My thinking above was if we were using the same endpoint for both, the API would need to perform some logic there.

Reply:
Perfect timing, @mbodeantor I've got almost everything else ready to go locally for the client stuff.

Reply:
@josh-chamberlain Looks like we will need an email server in order to properly send these password reset emails. Do we already have one by chance? If not, looks like Sendgrid has a limited free plan:
https://sendgrid.com/en-us/marketing/sendgrid-services-cro#pricing

Reply:
@mbodeantor @josh-chamberlain Doesn't python have built-in support for sending smtp emails? Presumably we already have email set up for pdap.io, couldn't we just set up a no-reply address and send using that? (Sendgrid is great, though. Have used previously for node projects, where there is no native email functionality.)

Reply:
@josh-chamberlain @joshuagraber An email server will allow our password reset emails to be successfully delivered, emails from a random python app will likely all be blocked